components/public-api-server/pkg/apiv1/workspace.go

```diff
@@ -76,6 +76,50 @@ func (s *WorkspaceService) GetWorkspace(ctx context.Context, req *connect.Reques
 	}), nil
 }
 
+func (s *WorkspaceService) WorkspaceStatusUpdate(ctx context.Context, req *connect.Request[v1.WorkspaceStatusUpdateRequest], stream *connect.ServerStream[v1.WorkspaceStatusUpdateResponse]) error {
+	logger := ctxlogrus.Extract(ctx)
+
+	conn, err := getConnection(ctx, s.connectionPool)
+	if err != nil {
+		return err
+	}
+
+	workspace, err := conn.GetWorkspace(ctx, req.Msg.GetWorkspaceId())
+	if err != nil {
+		logger.WithError(err).Error("Failed to get workspace.")
+		return proxy.ConvertError(err)
+	}
+
+	if workspace.LatestInstance == nil {
+		logger.WithError(err).Error("Failed to get latest instance.")
+		return connect.NewError(connect.CodeFailedPrecondition, fmt.Errorf("instance not found"))
+	}
+
+	ch, err := conn.InstanceUpdates(ctx, workspace.LatestInstance.ID)
+	if err != nil {
+		logger.WithError(err).Error("Failed to get workspace instance updates.")
+		return proxy.ConvertError(err)
+	}
+
+	for {
+		select {
+		case update := <-ch:
+			instance, err := convertWorkspaceInstance(update, workspace.Workspace.Shareable)
+			if err != nil {
+				logger.WithError(err).Error("Failed to convert workspace instance.")
+				return proxy.ConvertError(err)
+			}
+			_ = stream.Send(&v1.WorkspaceStatusUpdateResponse{
+				Result: &v1.WorkspaceStatus{
+					Instance: instance,
+				},
+			})
+		case <-ctx.Done():
+			return nil
+		}
```

Comment:
You need to close `ch`, otherwise we'll continue to receive updates on the channel which will leak

Reply:
Or somehow you need to stop listening

Reply:
I don't think is possible to close readonly channels, or am I misunderstanding :thinking:

Reply:
Ah, the channel is owned (and managed) by the `APIOverJSON` struct. It does it using 
```

	go func() {
		<-ctx.Done()

		gp.mu.Lock()
		delete(gp.subs[instanceID], chn)
		close(chn)
		gp.mu.Unlock()
	}()
```

This means that the `ctx` is used to control how long that channel remains open, and when it's closed.

This also means that our implementation no longer needs to check for `ctx.Done()`, it's done downstream. So we should consume the updates simply with 
```
for update := range ch {
	...
}
```
This will ensure graceful termination once we read all the items from the channel (and the chan becomes closed).